src/bans.rs

```diff
@@ -0,0 +1,322 @@
+use arc_swap::ArcSwap;
+use log::error;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::config::get_ban_time;
+use crate::config::Address;
+use crate::pool::PoolIdentifier;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::TrySendError;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::Duration;
+use tokio::time::Instant;
+
+pub type BanList = HashMap<PoolIdentifier, HashMap<Address, BanEntry>>;
+
+#[derive(Debug, Clone, Copy)]
+pub enum BanReason {
+    FailedHealthCheck,
+    MessageSendFailed,
+    MessageReceiveFailed,
+    FailedCheckout,
+    StatementTimeout,
+    #[allow(dead_code)]
+    ManualBan,
+}
+#[derive(Debug, Clone)]
+pub struct BanEntry {
+    reason: BanReason,
+    time: Instant,
+    duration: Duration,
+}
+
+#[derive(Debug, Clone)]
+pub enum BanEvent {
+    Ban {
+        pool_id: PoolIdentifier,
+        address: Address,
+        reason: BanReason,
+    },
+    Unban {
+        pool_id: PoolIdentifier,
+        address: Address,
+    },
+    CleanUpBanList,
+}
+
+impl BanEntry {
+    pub fn has_expired(&self) -> bool {
+        return Instant::now().duration_since(self.time) > self.duration;
+    }
+
+    pub fn is_active(&self) -> bool {
+        !self.has_expired()
+    }
+}
+static BANLIST: Lazy<ArcSwap<BanList>> = Lazy::new(|| ArcSwap::from_pointee(BanList::default()));
+
+static BAN_MANAGER: Lazy<ArcSwap<BanManager>> =
+    Lazy::new(|| ArcSwap::from_pointee(BanManager::default()));
+
+#[derive(Clone, Debug)]
+pub struct BanManager {
+    channel_to_worker: Sender<BanEvent>,
+}
+
+impl Default for BanManager {
+    fn default() -> BanManager {
+        let (channel_to_worker, _rx) = channel(1000);
```

Comment:
Thinking out loud:

This channel can get saturated during incidents. Imagine we have 30k clients that together produce 50k QPS. If a replica goes down, all of a sudden we'll get 50k banning events which will quickly saturate this channel. The async worker task may not even get scheduled to ban the replica.

With an RW lock, one of the clients will ban, and all the others will read the banlist and see that it's banned immediately. RW locks are probably more expensive than ArcSwap (something we should actually validate), but they are more effective than channels I think (something we should investigate).

I think the middle ground can be a Mutex for banning, and an ArcSwap for reading the list, so only one task gets to ban at a time instead of a thundering herd (pseudo-code):

```rust
fn ban() {
  let guard = ban_mutex.lock();

  let new_ban_list = ...;

  arc_swap.set(new_ban_list);
}

fn is_banned() {
   let ban_list = (*arc_swap);
   ban_list.contains_key(address);
}
```

The thesis that this architecture allows for other agents than clients to ban seems shaky. Another agent, e.g. an admin command, can easily use the `pool.ban` and `pool.unban` methods, for example, and it would require very little changes to the code and no changes to the arch.